src/Application.vala

```diff
@@ -48,8 +48,63 @@ public class Mail.Application : Gtk.Application {
         set_accels_for_action ("app.quit", {"<Control>q"});
     }
 
-    public override void open (File[] files, string hint) {
-        activate ();
+    public override int command_line (ApplicationCommandLine command_line) {
+        activate();
+
+        string[] argv = command_line.get_arguments ();
+
+        // The only arguments we support are mailto: URLs passed in by the OS. See RFC 2368 for
+        // details. We handle the most commonly used fields.
+        foreach (var mailto_uri in argv[1:argv.length]) {
+            string to = null, cc = null, subject = null;
+
+            try {
+                Soup.URI mailto = new Soup.URI(mailto_uri);
+                if (mailto == null) {
+                    throw new OptionError.BAD_VALUE ("Argument is not a URL.");
+                }
+                if (mailto.scheme != "mailto") {
+                    throw new OptionError.BAD_VALUE ("Cannot open non-mailto: URL");
+                }
+
+                to = Soup.URI.decode(mailto.path);
+                if (to == null || to == "") {
+                    throw new OptionError.BAD_VALUE ("mailto: URL does not specify a recipent");
+                }
+                if (mailto.query != null) {
+                    var params = parse_mailto_query (mailto.query);
+                    if (params["cc"] != null) {
+                        cc = params["cc"];
+                    }
+                    if (params["subject"] != null) {
+                        subject = params["subject"];
+                    }
+                }
+
+                new ComposerWindow.with_headers (main_window, to, cc, subject).show_all ();
+            }
+            catch (OptionError e) {
+                warning ("Argument parsing error. %s", e.message);
+            }
+        }
+
+        return 0;
+    }
+
+    private Gee.HashMap<string, string> parse_mailto_query (string query) throws OptionError {
+        var result = new Gee.HashMap<string, string> ();
+        var params = query.split ("&");
+
+        foreach (string param in params) {
```

Comment:
```suggestion
        foreach (unowned string param in params) {
```

Reply:
I'll ask about this in #vala, because I think I understand why this is necessary now, but it seems subtle.